src/main/java/com/spotify/confidence/ConfidenceValue.java

```diff
@@ -0,0 +1,547 @@
+package com.spotify.confidence;
+
+import com.google.common.annotations.Beta;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import com.google.protobuf.ListValue;
+import com.google.protobuf.NullValue;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Beta
+public abstract class ConfidenceValue {
+
+  static final ConfidenceValue NULL_VALUE =
+      new ConfidenceValue() {
+
+        @Override
+        public boolean isNull() {
+          return true;
+        }
+
+        @Override
+        public com.google.protobuf.Value toProto() {
+          return com.google.protobuf.Value.newBuilder().setNullValue(NullValue.NULL_VALUE).build();
+        }
+      };
+
+  private ConfidenceValue() {}
+
+  public boolean isStruct() {
+    return false;
+  }
+
+  public boolean isBoolean() {
+    return false;
+  }
+
+  public boolean isString() {
+    return false;
+  }
+
+  public boolean isInteger() {
+    return false;
+  }
+
+  public boolean isDouble() {
+    return false;
+  }
+
+  public boolean isTimestamp() {
+    return false;
+  }
+
+  public boolean isDate() {
+    return false;
+  }
+
+  public boolean isNull() {
+    return false;
+  }
+
+  public boolean isList() {
+    return false;
+  }
+
+  public Struct asStruct() {
+    throw new IllegalStateException("Not a StructValue");
+  }
+
+  public String asString() {
+    throw new IllegalStateException("Not a StringValue");
+  }
+
+  public int asInteger() {
+    throw new IllegalStateException("Not a IntegerValue");
+  }
+
+  public double asDouble() {
+    throw new IllegalStateException("Not a DoubleValue");
+  }
+
+  public Instant asInstant() {
+    throw new IllegalStateException("Not a InstantValue");
+  }
+
+  public LocalDate asLocalDate() {
+    throw new IllegalStateException("Not a DateValue");
+  }
+
+  public boolean asBoolean() {
+    throw new IllegalStateException("Not a BooleanValue");
+  }
+
+  public java.util.List<ConfidenceValue> asList() {
+    throw new IllegalStateException("Not a ListValue");
+  }
+
+  public static ConfidenceValue of(int value) {
```

Comment:
It's a bit strange that these `of(<primitive>)` creation functions return a ConfidenceValue whereas the lists and structs doesn't. Maybe we should actually create a Primitive type?

Reply:
Not a blocker though, but food for thought.